core/src/main/java/com/amplifyframework/core/model/AuthRule.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.core.model;
+
+import androidx.core.util.ObjectsCompat;
+
+import java.util.Arrays;
+
+/**
+ * {@link AuthRule} is used define an authorization rule for who can access and operate against a
+ * {@link Model} or a {@link ModelField}.
+ *
+ * @see <a href="https://docs.amplify.aws/cli/graphql-transformer/directives#auth")>GraphQL Transformer @auth directive
+ * documentation.</a>
+ */
+public final class AuthRule {
```

Comment:
This class seems like it is "modal."  Like, depending on the `AuthStrategy`, the `AuthRule` will have different fields become null, by design, right?

Instead, would it make sense to have an `interface AuthRule`, and then do `GroupBasedAuthRule`, which only returned group info, `UserBasedAuthRule`, which only returned user info, etc.?

Reply:
You're right, it is somewhat "modal".   I did consider a similar approach to what you are a suggesting, but did not implement because I wasn't quite sure how I would handle an invalid rule definition.

For example if a customer defines a rule of: `{ allow: owner, groupsField: "groups" }`, (`groupsField` should only be used when `allow: groups`), what should happen?

Interestingly, `amplify push` does *not* fail with a schema like this. 

We could:
1. Annotate requests with a mismatch of rule metadata, and perhaps end up with an error in the `GraphQLResponse`.
2. Silently ignore the groupsField at runtime.
3. Throw a `RuntimeException` and suggest to the customer to fix their schema.

Anyway, I *think* refactoring to a "modal" design could make sense, but would like to address that in my next PR which will annotate requests with these details, after I've had a chance to dive deeper on how this really works.

Reply:
You could even make an `InvalidAuthRule` subclass!